enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,511 @@
+ - so two or more pods cannot access the same file
+ - but if each pod has its own file, its own copy, of the shared data, things work easily from a SELinux 
+ perspective
+ - this is where using ephemeral CSI storage comes in
+ - using ephemeral based storage allows for use of a tmpfs (virtual memory filesystem) for storage
+ - tmpfs is fast and has negligible overhead
+ - it will have to reconstructed if the Pod moves
+ - but how it eases dealing with SELinux protection is the bonus we are taking
+ - so each pod gets its own tmpfs
+ - the kubelet leverages tmpfs in a similar fashion already
+
+#### End to end flow (from creation, to opt-in, to validation, and then injections)
+
+An admin creates a new cluster level custom resource for encapsulating the sharable `Secret` or `ConfigMap`.
+
+
+```yaml
+kind: ProjectedResource
+apiVersion: projected-resource.storage.openshift.io/v1
+metadata:
+  name: shared-cool-secret
+spec:
+  backingResource:
+    kind: Secret
+    apiVersion: v1
+    namespace: ns-one
+    name: cool-secret
+  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
+status:
+  conditions:
+  ...
+```
+
+And sets up ACL/RBAC to control who can list/get such objects.
+
+```yaml
+rules:
+- apiGroups:
+  - projected-resource.storage.openshift.io
+  resources:
+  - projectedResources
+  resourceNames:
+  - shared-cool-secret
+  verbs:
+  - get
+  - list
+  - watch
+
+```
+
+A user queries this new type to see what is available to her/him.
+
+Then, to opt in, a user created Pod can cite a CSI volume references (see below for a more detailed explanation of them) that 
+references our new type:
+
+```yaml
+csi:
+  driver: projectedresources.storage.openshift.io
+    volumeAttributes:
+      projectedResourceName: the-projected-resource
+
+```
+
+For pre-existing objects that encapsulate a Pod, a new, documented annotation can be declared to facilitate injection
+of the CSI driver volume specification 
+
+A new admission plugin introduced as part of this feature can then:
+- mutate Pods with this annotation, adding the CSI driver volume specification, if the Pod `ServiceAccount` is allowed sufficient 
+access the `ProjectedResource` (via a SAR check)
+- if a Pod already happens to have the CSI Volume specified with reference to a `ProjectedResource`, again see if the 
+associated `ServiceAccount` for the Pod has sufficient access to the `ProjectedResource`
+- if sufficient access does not exist, the admission plugin marks the Pod as invalid/forbidden.
+- if sufficient access does exists, the `volumeAttributes` are also updated to add a key/value pair for the requesting
+`ServiceAccont` (including both its name and namespace)
+
+```yaml
+csi:
+  driver: projectedresources.storage.openshift.io
+    volumeAttributes:
+      projectedResourceName: the-projected-resource
+      requestingNamespace: the-users-namespace
+      requestingServiceAccount: the-sa-the-user-is-using-for-this-object
```

Comment:
Why is namespace / service account necessary? Admission plugin can see it in the Pod spec it is admitting and CSI driver can get it in `NodePublishVolume` call.

Reply:
it was the ability of the "CSI driver can get it in `NodePublishVolume` call" that I had not discovered to date

on your prompting, I revisited secrets-store-csi-driver and sure enough I see 

```
		parameters["csi.storage.k8s.io/pod.name"] = attrib["csi.storage.k8s.io/pod.name"]
		parameters["csi.storage.k8s.io/pod.namespace"] = attrib["csi.storage.k8s.io/pod.namespace"]
```

where `attrib` comes from `NodePublishVolumeRequest.GetVolumeContext()`

I'm also finding those strings in vendored k8s code, as well as `csi.storage.k8s.io/serviceAccount.name`

I'll update accordingly, thanks

Reply:
Yes, that's how the CSI driver gets it. `CSIDriver.spec.podInfoOnMount: true` must be set to enable this behavior on Kubernetes side. See a sample CSIDriver deployment : https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/deploy/kubernetes-1.17/hostpath/csi-hostpath-driverinfo.yaml#L12

Reply:
cool ... adding that note